service/history/retry.go

```diff
@@ -93,18 +93,24 @@ func isRetryable(failure *failurepb.Failure, nonRetryableTypes []string) bool {
 		return false
 	}
 
-	if failure.GetApplicationFailureInfo().GetNonRetryable() || failure.GetServerFailureInfo().GetNonRetryable() {
-		return false
-	}
-
 	if failure.GetTimeoutFailureInfo() != nil {
 		if failure.GetTimeoutFailureInfo().GetTimeoutType() != enumspb.TIMEOUT_TYPE_START_TO_CLOSE &&
 			failure.GetTimeoutFailureInfo().GetTimeoutType() != enumspb.TIMEOUT_TYPE_HEARTBEAT {
 			return false
+		} else {
+			return true
 		}
 	}
 
+	if failure.GetServerFailureInfo() != nil {
+		return !failure.GetServerFailureInfo().GetNonRetryable()
+	}
+
 	if failure.GetApplicationFailureInfo() != nil {
+		if failure.GetApplicationFailureInfo().GetNonRetryable() {
```

Comment:
[Nit] why the different style between line 106 (returning !) vs here (if true then return false)?

Reply:
`ServerFailure` doesn't have `type` field. Therefore retryable is controlled only by `NonRetryable` flag (if it is `true`, `IsRetryable` is `false` and vice versa). `ApplicationFailure` also has a `type` field which needs to be checked against `nonRetryableTypes`. So if `NonRetryable` is set we can return right away. If not, we need to check type.